### PR TITLE
Make `placeholder` of DbApiHook configurable in UI

### PR DESCRIPTION
**Pull request by dabla**
Moved the placeholder property logic from OdbcHook to DbApiHook.
We now encountered the same issue when using jdbc connections as we did back then in odbc connections where the placeholder should be ? instead of %.  As OdbcHook and JdbcHook both inherit DbApiHook, I moved the property there so both classes can use same logic.  Also moved test of placeholder property from OdbcHook to DbApiHook and did some refactoring so both tests can reuse same mocking function.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: tests/providers/conftest.py [hunk 1]
```diff
@@ -55,3 +59,32 @@ def hook_conn(request):
             )
 
         yield m
+
+
+def mock_hook(hook_class: type[BaseHook], hook_params=None, conn_params=None):
```
Comment (Taragolis) [inline tests/providers/conftest.py]: `conftest.py` it not a good place for store importable stuff, it intends to use for configure tests (some things limited by root conftest), define fixtures and some other stuff which accessible in `pytest` runtime

Comment (dabla) [reply tests/providers/conftest.py]: So I just put it in another file then?  Because I've cleaned up a bit the test and made a reusable mock across odbc and jdbc tests so we have less duplicate code

Comment (dabla) [reply tests/providers/conftest.py]: I moved it to test_utils under common sql

File: airflow/providers/common/sql/hooks/sql.py [hunk 2]
```diff
@@ -50,6 +50,7 @@
     from airflow.providers.openlineage.sqlparser import DatabaseInfo
 
 T = TypeVar("T")
+DEFAULT_SQL_PLACEHOLDERS = frozenset({"%s", "?"})
```
Comment (eladkal) [inline airflow/providers/common/sql/hooks/sql.py]:
```suggestion
SQL_PLACEHOLDERS = frozenset({"%s", "?"})
```

It's not defaults it's allowed placeholders

File: airflow/providers/common/sql/hooks/sql.py [hunk 3]
```diff
@@ -173,7 +175,17 @@ def __init__(self, *args, schema: str | None = None, log_sql: bool = True, **kwa
...
-    def placeholder(self) -> str:
+    def placeholder(self):
+        conn = self.get_connection(getattr(self, self.conn_name_attr))
+        placeholder = conn.extra_dejson.get("placeholder")
+        if placeholder in DEFAULT_SQL_PLACEHOLDERS:
+            return placeholder
+        self.log.warning(
+            "Placeholder defined in Connection '%s' is not listed in 'DEFAULT_SQL_PLACEHOLDERS' "
+            "and got ignored. Falling back to the default placeholder '%s'.",
+            placeholder,
```
Comment (eladkal) [inline airflow/providers/common/sql/hooks/sql.py:186]: I wonder if this is right approach?
I would prefer to warn when using placeholder that is different with what we know but if user still wish to use something that is different we should allow it and not prevent it.

Comment (dabla) [reply airflow/providers/common/sql/hooks/sql.py:186]: We can change it if you want, but that's how it was originally implemented in the OdbcHook.

**Review by potiuk — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Comment by dabla** [comment]
Thank you for merging this pull request, could someone have a look to check this one also please: https://github.com/apache/airflow/pull/38707
